### PR TITLE
Stop config-supplied values reaching exec, eval and getattr sinks

### PR DESCRIPTION
**Pull request by danielhanchen**
### Summary

Follow-up to #1083. That one closed a real arbitrary-code-execution path: `model_type`, read straight out of a downloaded `config.json`, was interpolated into a string that was then `exec`'d, so a hostile model repo got code execution with no `trust_remote_code`. It validated at the producer, `get_transformers_model_type`.

This is the audit of the remaining `exec` / `eval` sites in the package. Most of them are fine and are left alone: the ~46 in `compiler.py` execute source this package generated itself, out of `inspect.getsource` of an installed transformers class or `dir()` of an installed module, and `vllm_utils.py`'s `exec(f"new_model.{layer_name} = layer")` takes its layer names from hardcoded templates in `get_model_layer_config`, not from checkpoint keys. What follows is what was left.

### `compiler.py`, defence in depth at the sink

`unsloth_compile_transformers` rebuilds `transformers.models.<model_type>.modeling_<model_type>`, and `create_new_function` `os.path.join`s a name into the compiled-cache filename, which is a path-traversal sink. Both take those values as parameters and trust that whoever called them went through `hf_utils`. Each now re-checks its own input, with the same regex and the same `ValueError` wording as the #1083 guard. This does not replace the producer-side check, it means the sink no longer depends on it.

The `UNSLOTH_FORCE_CUSTOM_DTYPE` dtype field also went through `eval`. It is one of `None`, `torch.float16`, `torch.bfloat16`, `torch.float32`, so it goes through `_get_dtype` now. `_get_dtype` returns `None` for anything it does not recognise, which is the same "no override" branch the `None` field already took.

### `empty_model.py`, architectures were an open `getattr`

`create_empty_model` did `getattr(transformers, config.architectures[0])` and called the result. `architectures` comes out of a downloaded `config.json`, and the attribute is not checked against anything, so a config could name any public symbol on the `transformers` namespace and have it constructed. The twin at `vllm_utils.py:3835` already has a gate, this brings the two in line.

It is now checked against the names transformers registers in its `MODEL_*_MAPPING_NAMES` tables. `CONFIG_MAPPING_NAMES` is deliberately excluded, it holds config classes rather than architectures. One thing worth flagging from testing: an `issubclass(PreTrainedModel)` gate, which was my first attempt, wrongly rejects entries like `HiggsAudioV2TokenizerModel` on transformers 5.15.1, because they are lazy placeholders for a missing optional dependency. Going through the registry avoids masking transformers' own error for those.

This also adds `_get_module_attribute` next to the `_set_module_attribute` that #1083 introduced, same numeric-segment handling and same raise-loudly semantics.

### `peft_utils.py`, `tokenizer_utils.py`, mechanical `eval` removal

Three `eval(f"...")` attribute lookups over parameter names and special-token names, replaced by that `getattr` walk. These are safe today, both feed off fixed tables, but there is no reason for a name that came out of a checkpoint to be parsed as a Python expression. I checked the walk against the `eval` it replaces on a real LoRA model: 72 paths resolved, 0 mismatches. Note `peft_utils.py:399` already rewrites `.0.` to `[0].` before splitting, so no numeric component actually reaches the resolver on that path.

### `saving_utils.py`

The output dtype was pinned by a bare `assert`, which disappears under `python -O`. It raises a `ValueError` now.

### Lint

`scripts/lint_dynamic_exec.py` walks the AST of every package file and fails on a call to `exec` / `eval` / `compile` whose first argument is an f-string, a string `BinOp`, or a `%` / `.format()` construction. That is the "untrusted value spliced into source" shape specifically, not bare `exec(source)`.

The 53 reviewed sites live in `scripts/dynamic_exec_allowlist.json`, each with a one-line reason, keyed on the path, the enclosing qualname and a sha256 of `ast.unparse` of the call. Reformatting a site keeps its justification, changing what gets interpolated invalidates it. Stdlib only, so the `source-lint` job needs no new dependency beyond the `ruff` and `pyyaml` it already installs, and it has a `--self-test` that runs first so a broken lint fails loudly rather than passing everything.

### Testing

- `tests/security` plus `tests/test_module_attribute_getter.py` on this branch: 114 passed
- full suite, `main` vs this branch: 1 failed / 5,236 passed before, 2 failed / 5,401 passed after. Both failures are in `test_rmsnorm_recompile_guards.py`, which passes in isolation on both trees and varied 1 to 2 across repeat runs of an essentially identical tree. It is order dependent on the shared `torch.compile` cache budget, not on this change
- version matrix, transformers 4.57.6 / 5.0.0 / 5.15.1 and PEFT 0.18.0 / 0.19.1 / 0.20.0: 0 model types wrongly rejected out of 4,586 to 5,768 checked per cell, 0 architectures wrongly rejected, 0 gadgets allowed, 115 PEFT paths with 0 mismatches
- the model-type sink test mirrors the #1083 `CONFIG_MAPPING_NAMES | dir(transformers.models)` sweep, so a legitimate type cannot regress silently

### Relationship to the other open PRs

#1106 covers the `quantize_gguf` command injection and #1105 covers the `create_locked_down_function` builtins leak, which were the other two findings in this audit. Both are left to those PRs and nothing here touches `llama_cpp.py` or `rl_environments.py`, so this should merge in any order relative to them. The unsloth-side half of the audit is a separate PR on that repo.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Ran a set of simulations against this branch rather than just the test suite, since every guard here converts a value that used to flow through into a `ValueError`, and each one is therefore a chance to break a model that works today.

### What the guards would reject

The question is not whether they block an attack, it is whether they block anything real. Swept across the supported transformers range:

| transformers | model_types | architectures | PreTrainedModel classes on the namespace | wrongly rejected |
|---|---|---|---|---|
| 4.57.6 | 421 | 1174 | 1725 | 0 |
| 5.0.0 | 444 | 1180 | - | 0 |
| 5.8.0 | 623 | 1299 | 1924 | 0 |
| 5.15.1 | 693 | 1421 | 2095 | 0 |

PEFT 0.18.0, 0.19.1 and 0.20.0 at transformers 5.15.1, same result.

Two things worth calling out from that sweep:

- 38 of the model types transformers ships contain a hyphen (`data2vec-audio`, `vision-encoder-decoder`, ...) and none of them match `[a-z0-9_]+` raw. They only pass because `get_transformers_model_type` rewrites `-` to `_` before its own check, so "always go through the producer" is load-bearing rather than incidental. The sink check is the same regex as the producer's, so the two cannot disagree.
- an `issubclass(PreTrainedModel)` gate, which was my first attempt at the architectures check, wrongly rejects entries like `HiggsAudioV2TokenizerModel` on 5.15.1, because they are lazy placeholders for a missing optional dependency. Going through the auto-mapping registry avoids masking transformers' own "install X" error. The subclass check is kept only as a fallback, which also covers a class registered after `_registered_architectures`' `lru_cache` was filled (verified both ways).

### What actually reaches the name guard

Rather than reasoning about it, I wrapped `create_new_function` and drove the real compiler over 45 model types. 572 names recorded, 56 unique, every one a valid identifier, none rejected. Longest was `unsloth_compiled_module_modernbert`.

### The module walk

`_get_module_attribute` checked against the `eval` it replaces as the oracle on a real model: 76 paths, 0 mismatches, and it additionally resolves the 66 paths with numeric segments that `eval` could not index at all. Unresolvable paths raise rather than returning None, and it cannot evaluate an expression (`layers[0]`, `1+1`, `__class__()` all raise). The seven `tokenizer.<x>_id` lookups match the old `eval` exactly.

### The assert that became a raise

This is the only behaviour change in the PR, and it only shows up under `python -O`:

| | old (`assert`) | new (`raise`) |
|---|---|---|
| without `-O` | AssertionError | ValueError |
| with `-O` | **allowed through** | ValueError |

Nothing in `unsloth_zoo` catches `AssertionError` anywhere, so the exception type change has no catcher to break.

### Platform

No changed line branches on OS or accelerator, checked against the merge base. Both guards reject Windows separators and traversal (`..\..\windows\system32`, `C:\Windows`, `\\server\share`). No generated module name collides with a DOS reserved device name, which matters because the name becomes a filename. No UI or Studio file is touched.

Tests on this branch: `tests/security` plus `tests/test_module_attribute_getter.py`, 114 passed.

**Comment by danielhanchen** [comment]
Ran `reviewer.py` over this PR (12 parallel reviewers). Three runtime defects, two of which I had introduced and one of which was an incorrect claim in my own comment. Each was reproduced before fixing.

### `empty_model.py`: a rejected architecture leaked global state (1/12)

`create_empty_vision_model` replaces `SiglipVisionModel._init_weights` with a no-op so meta-device construction skips initialisation, and restores it at the end of the function. My guard raised in between. A config with an unknown architecture therefore left weight init disabled on a class shared by the entire process, and every SigLIP model built afterwards silently skipped initialisation. Silent, and nowhere near the cause.

Reproduced the old ordering directly to confirm the leak was real rather than theoretical, so the regression test is not vacuous:

```
PRE-FIX leaked _init_weights: True
PRE-FIX left sentinel: True
```

The guard now runs before the patch, and a test pins that ordering in the source so it cannot drift back.

### `empty_model.py`: the getattr walk could not resolve bracket components (1/12)

`peft_utils` rewrites `.0.` to `[0].` before splitting. Its first call site skips paths containing brackets, but the second does not: when a module's forward calls `self.blocks[0](...)` directly, `final_where` includes the bracketed component and `getattr(obj, "blocks[0]")` fails where the removed `eval` indexed it happily. That is gradient checkpointing breaking on a real model shape.

I had previously checked this and concluded no bracket component could reach the resolver. That was right for the first call site and wrong for the second. The walk now resolves digit subscripts, and only digits, so it stays a name-and-index walk rather than an expression: `blocks[0+0]`, `blocks[x]`, `blocks[-1]` and `blocks['a']` all still raise.

### `vllm_utils.py`: the architecture guard was asymmetric (6/12)

The vLLM vision path does the same `getattr(transformers, config.architectures[0])` followed by `from_pretrained`, gated only on `model_type`. `model_type` and `architectures` are independent fields of the same downloaded config.json, so that gate constrains neither. A config declaring a supported `model_type` and `architectures: ["AutoTokenizer"]` resolved and invoked an unrelated class.

My comment in `empty_model.py` claimed this path already applied the same gate. **That claim was false.** It is removed rather than reworded, and the gate is now actually applied there.

### Lint

Same four corrections as unslothai/unsloth#9777: identity is `path::qualname::hash`, locals bound to a built string are tracked, `builtins.exec` is a sink, rebinding clears. The committed allowlist had 53 entries collapsing to 40 keys, so approvals were genuinely being shared across unrelated locations here too.

The stronger lint found four more interpolated sites, all the variable form and all self-generated from installed packages: the gguf reader partial class, the peft lora submodule scan, and the transformers quantizer patch. Recorded with reasons. 53 to 57.

### Noted, not changed here

`mlx/loader.py:1297,1682` (`processor_class`, `image_processor_type` from downloaded JSON) and `compiler.py:1658` (`create_standalone_class` evaluates its `model_location` parameter, bypassing the sink guard when called directly) are the same config-to-constructor shape but different sinks, and neither is touched by this PR. They belong in a follow-up rather than being folded in here.

### Verification

`tests/security` plus `tests/test_module_attribute_getter.py`: 118 passed. Guard sweeps re-run clean across transformers 4.57.6 / 5.0.0 / 5.8.0 / 5.15.1 and PEFT 0.18.0 / 0.19.1 / 0.20.0, still 0 wrongly rejected.

File: scripts/lint_dynamic_exec.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+def load_allowlist() -> dict[str, dict]:
+    if not ALLOWLIST_PATH.exists():
+        return {}
+    data = json.loads(ALLOWLIST_PATH.read_text())
+    return {entry["hash"]: entry for entry in data.get("allowed", [])}
+
+
+def key_of(finding: dict) -> str:
+    return finding["hash"]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply scripts/lint_dynamic_exec.py]: Fixed in 25d6d91: key_of is path::qualname::hash now. This repo's allowlist had 53 entries collapsing to 40 keys, so approvals really were shared across unrelated locations.

File: scripts/lint_dynamic_exec.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+    visit_FunctionDef = _enter
+    visit_AsyncFunctionDef = _enter
+    visit_ClassDef = _enter
+
+    def visit_Call(self, node: ast.Call):
+        function = node.func
+        if isinstance(function, ast.Name) and function.id in SINKS and node.args:
+            reason = _is_interpolated(node.args[0])
+            if reason is not None:
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply scripts/lint_dynamic_exec.py]: Fixed in 25d6d91: locals bound to a built string are tracked per scope, and rebinding to a non-interpolated value clears the name.

File: unsloth_zoo/empty_model.py [hunk 3]
```diff
@@ -490,7 +491,17 @@ def _init_weights(self, module):
...
+    # `architectures[0]` is a raw string off the downloaded config.json, and `getattr`
+    # on the transformers namespace followed by a call is an unbounded gadget: nothing
+    # here requires it to name a model class. Restrict it to the architectures the auto
+    # mappings actually register, which is the same gate the vllm_utils.py twin applies.
+    architecture = config.architectures[0]
+    if not _is_known_architecture(architecture):
+        raise ValueError(
+            f"Unsloth: config.json declares architecture `{architecture}`, which is not "
+            f"a model architecture transformers registers."
+        )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/empty_model.py:499]: Fixed: the guard now runs BEFORE the patch rather than being unwound after it, and the restore moved into a finally gated on whether this call installed the patch. Reproducing the old ordering confirmed the leak was real (restored=False, sentinel=True), including under a KeyboardInterrupt, which except Exception never caught.

File: scripts/dynamic_exec_allowlist.json [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+      "hash": "d9650e96220991c3",
+      "reason": "`norm` / `block` come from dir(timm.layers.norm_act) and a class regex over inspect.getsource(timm.models._efficientnet_blocks), i.e. names in the installed timm."
+    },
+    {
+      "path": "unsloth_zoo/compiler.py",
+      "qualname": "create_standalone_class",
+      "sink": "eval",
+      "kind": "f-string",
+      "hash": "41335056252b5222",
+      "reason": "`module` is a class name every caller takes from dir(modeling_file) or a `class X(...)` regex over inspect.getsource of the installed modeling file."
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply scripts/dynamic_exec_allowlist.json]: The justification was incomplete and I have rewritten it in 25d6d91, but not adding a guard. create_standalone_class is not exported: it is absent from compiler.__all__ (5 names), and all five call sites are in the same file passing f"transformers.models.{model_type}.modeling_{model_type}" built from a model_type get_transformers_model_type already validated. There is no supported path that reaches it with a config-derived value.

File: scripts/lint_dynamic_exec.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+        version: moving the interpolation onto the line above is a two-line refactor
+        that leaves the vulnerability shape exactly as it was. Rebinding a tracked name
+        to something that is not a built string clears it, so this does not accumulate
+        false positives.
+        """
+        reason = _is_interpolated(node.value)
+        for target in node.targets:
+            if not isinstance(target, ast.Name): continue
+            if reason is None:
+                self.tainted[-1].pop(target.id, None)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply scripts/lint_dynamic_exec.py:153]: Not taking this one. Branch-sensitive taint is a dataflow analysis and this is a heuristic CI gate. One level of local indirection is fixed because that is a two-line refactor of an existing shape.

File: scripts/lint_dynamic_exec.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+def key_of(finding: dict) -> str:
+    """Identity of a reviewed call: where it is, and what it is.
+
+    The hash alone is not an identity. Two different call sites with byte-identical
+    source share a hash, so approving one approved the other, and copying an
+    allowlisted call into a function with attacker-controlled inputs inherited its
+    justification without review. The location is half of what was reviewed, so it is
+    half of the key.
+    """
+    return f"{finding['path']}::{finding['qualname']}::{finding['hash']}"
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply scripts/lint_dynamic_exec.py:254]: Not taking this one. Two identical calls in the same function legitimately share one justification, and adding a positional discriminator would reintroduce the line-number drift the ast.unparse hash exists to avoid. The property that matters, no approval spanning two different locations, holds with path::qualname::hash.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Pushed `891a9d2`. This PR had already converged on `25d6d910`, so re-opening it is a deliberate choice rather than an oversight.

All three repos carry the same `scripts/lint_dynamic_exec.py` apart from `DEFAULT_TARGETS`. Review on the notebooks and unsloth copies turned up four genuine bypasses of the interpolation check, and leaving them fixed in two repos and open in this one would mean shipping a checker that is weaker here than the identical file elsewhere. Carrying the fix is worth losing the approval on the previous head.

The four, all confirmed to actually execute on CPython 3.13:

- `exec("import {module}".format_map(cfg))`, since `format_map` was not in the builder list
- `payload: str = f"import {name}"` then `exec(payload)`, since `ast.AnnAssign` is a sibling of `ast.Assign` and `visit_Assign` never saw it
- `exec(f"import {name}".encode())`, since `exec` accepts bytes and not only str
- 500 nested `not` in front of an interpolated `exec`, which exhausted the walk and was reported clean

The last one reverses an earlier decision of mine. I had made `RecursionError` return `[]` so a pathological file could not take the gate down, but this gate reads incoming PR contents, so a condition an attacker can trigger on demand must never be the thing that silences the check. It now raises a `ScanError` counted as `unscannable` in the FAIL line, which fails the gate while still printing every other finding.

No runtime code in this repo changed. The diff is `scripts/lint_dynamic_exec.py` and its test only.

Verification on this branch: allowlist count unchanged at **57**, so the new detections match nothing already in tree and no existing entry needed re-justification. Self test and full scan green. `tests/security` 95 passed. Full suite 5341 passed, 200 skipped, 0 failed.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Correction to the comment above: this PR was already merged at 10:31 when I posted it, so it was not re-opened and `891a9d2` never landed here. I pushed to the branch after the squash merge had closed it, and read the stale head as if the PR were still open.

The lint fix is now #1111, cherry-picked onto current `main`. Everything in this PR merged as intended; only the follow-up was orphaned.
